### ISSUE TITLE
Is it desirable to extend LocalPathStrategy?

### ISSUE THREAD

**Issue by mikrosk** [commonsguy/cwac-provider#15]
I've finally managed to align my code with #11. I'd like to provide an `insert()` for my files so I've been thinking to just extend `LocalPathStrategy` and add this function. However, without `getFileForUri()` being `protected` it's very inconvenient. I'd rather avoid copying & pasting your code to a new stream strategy.

Same goes for `StreamProvider`'s `buildPath()` but this method is not that long/complex so I could live with a small copied&pasted code at this occasion.

Also, it's not clear to me how I am supposed to work with `getValueForQueryColumn()`? If I have a new strategy, I will "register" it in `buildStrategy()` and then what? How can I access my (extended) strategy which is accessible through new columns? Should I make a private reference to it in `buildStrategy()`? (in this case the original strategies can't work with custom columns even if they could -- imagine a `getUniqueFilenameForThisUri()`).

**Comment by commonsguy** [comment]
With regards to your first two paragraphs, are you asking me to make those methods `protected`, or are  you asking something else? I am fine with making those `protected`, but I want to make sure that is what you are seeking.

> Also, it's not clear to me how I am supposed to work with `getValueForQueryColumn()`? 

Most likely, you do nothing with it.

> How can I access my (extended) strategy which is accessible through new columns?

That's not something that I have envisioned anyone doing. Mostly, the `getValueForQueryColumn()` hook exists to try to patch other problems in buggy clients, returning values to try to make those clients handle your streams despite their bugs. This is what the `_DATA` and `MIME_TYPE` columns are for. This was not envisioned as being some general-purpose mechanism for passing data between two apps that you are writing, which I assume is the case.

If you have further questions, please ask questions on Stack Overflow, as is covered in [the documentation](https://github.com/commonsguy/cwac-provider#questions).

**Comment by mikrosk** [comment]
> With regards to your first two paragraphs, are you asking me to make those methods protected, or are you asking something else? I am fine with making those protected, but I want to make sure that is what you are seeking.
That's exactly what am I seeking. :)

About `getValueForQueryColumn()`: well, in my case, I wanted to make file's modification time accessible via a new column. So it's not a patch, just an extension. If you would like to continue this on SO, I'll post the question there.

**Comment by commonsguy** [comment]
> That's exactly what am I seeking. :)

I will make that change at some point, though it might be next weekend before I get around to it.

>I wanted to make file's modification time accessible via a new column

Have a static `getFileForUri()` implementation that both your `StreamProvider` subclass and your `LocalFileStrategy` subclass share.

**Comment by mikrosk** [comment]
Thank you, I really appreciate it!

I like the idea of `getFileForUri()` being static, would it be too much to ask to make it `protected static` in your `LocalPathStrategy` as you are willing to change it to `protected` anyway?

**Comment by commonsguy** [comment]
> would it be too much to ask to make it protected static in your LocalPathStrategy

It won't work. My code would always call the `LocalPathStrategy` implementation, as it knows nothing of your subclass. You can't really override `static` methods the way you do instance ones.

**Comment by mikrosk** [comment]
But that's totally fine with me -- I just want to have this method accessible from both of my `LocalPathStrategy` & `StreamProvider` subclasses. Something as:

```
@Override
    protected Object getValueForQueryColumn(Uri uri, String col) {
        Object result;

        if (LAST_MODIFIED.equals(col)) {
            // TODO error handling
            result = LocalPathStrategy.getFileForUri(uri).getLastModified(uri); // calling your LocalPathStrategy here
        } else {
            result = super.getValueForQueryColumn(uri, col);
        }

        return result;
    }
```

and when overriding `insert()` and/or `update()`:

```
public class MyLocalPathStrategy extends LocalPathStrategy {
@Override
    public int update(Uri uri, ContentValues values, String selection, String[] selectionArgs) {
        File file = getFileForUri(uri); // calling your method again
       // use 'file' here...
    }
}
```

I don't need to override `getFileForUri()`, just making it `static` (for access from `StreamProvider` subclass) and `protected` (for `LocalPathStrategy` subclass).

Or, as mentioned, I would be totally happy to have just something as `getStrategy()` in `StreamProvider`, then I wouldn't need static `getFileForUri()`.

**Comment by commonsguy** [comment]
`getUriForFile()` is already `public` on `StreamProvider`.

I have made these changes:

- `buildPath()` is now `public` on `StreamProvider`
- `StreamProvider` now has a `protected` `getRootStrategy()` method, which returns the `CompositeStreamStrategy` used to aggregate all the individual strategies
- the `getStrategy()` method on `CompositeStreamStrategy` is now `public`

The latter two would be how you would handle getting at your strategy, given a `Uri` that resolves to that strategy.

**Comment by commonsguy** [comment]
Released in v0.4.2

**Comment by mikrosk** [comment]
`getUriForFile()` is public but `getFileForUri()` is not (and I need that one).

**Comment by commonsguy** [comment]
Ah, oops.

**Comment by commonsguy** [comment]
Addressed in v0.4.3.
